nano-sync/src/gadgets/mnt4/check_sig.rs

```diff
@@ -0,0 +1,91 @@
+use algebra::mnt4_753::Fr as MNT4Fr;
+use r1cs_core::SynthesisError;
+use r1cs_std::mnt6_753::{G1Gadget, G1PreparedGadget, G2Gadget, G2PreparedGadget, PairingGadget};
+use r1cs_std::{eq::EqGadget, fields::FieldGadget, pairing::PairingGadget as PG};
+
+/// This gadget checks that a given BLS signature is correct. But it only works if given the hash
+/// point, it does not have functionality to transform a message into the hash point.
+pub struct CheckSigGadget;
+
+impl CheckSigGadget {
+    /// This function checks the correctness of an aggregated signature for n different messages
+    /// from n different public keys. It implements the signature aggregation scheme from:
+    /// https://crypto.stanford.edu/%7Edabo/pubs/papers/aggreg.pdf
+    /// It implements the following verification formula:
+    /// e(sig, gen) = e(hash_1, pk_1) * e(hash_2, pk_2) * ... * e(hash_n, pk_n)
+    pub fn check_signatures<CS: r1cs_core::ConstraintSystem<MNT4Fr>>(
+        mut cs: CS,
+        public_keys: Vec<&G2Gadget>,
+        hash_points: Vec<&G1Gadget>,
+        signature: &G1Gadget,
+        generator: &G2Gadget,
+    ) -> Result<(), SynthesisError> {
+        // Make some initial sanity checks.
+        assert_eq!(
+            hash_points.len(),
+            public_keys.len(),
+            "One hash point per public key is required"
+        );
+
+        assert!(hash_points.len() > 1, "Min one message is required");
+
+        // Prepare all the public key elliptic curve points.
+        let mut pub_key_p_vars = vec![];
+
+        for (i, public_key) in public_keys.iter().enumerate() {
+            let pub_key_p_var: G2PreparedGadget =
+                PairingGadget::prepare_g2(cs.ns(|| format!("pub_key_p {}", i)), &public_key)?;
```

Comment:
```suggestion
                PairingGadget::prepare_g2(cs.ns(|| format!("pub_key_p {}", i)), public_key)?;
```
`public_key` should already be a reference.